### PR TITLE
Add support for cross-thread stack dumping to vespalib

### PR DESCRIPTION
**Pull request by vekterli**
@havardpe please review

There's no existing API for conveniently getting the stack trace of any other thread
than your own, so we have to turn things around a bit and make the target thread
dump its own stack in a way that we can then safely access and return it.

We allow for hooking `SIGUSR2` at process startup time and use this as our dedicated
stack dump signal internally. The process of stack dumping is then basically:

 1. Caller sends `SIGUSR2` to the target thread. This happens within a global mutex
    that ensures no other callers can attempt to dump stack at the same time. This
    is because we use a single, globally shared state between caller and target
    (the same signal handler function is used by all threads).
 2. Caller acquire-polls (with 1ms sleep) for target thread signal handler completion.
    Fancy technologies such as mutexes and condition variables are not safe to use
    from within a signal handler and therefore cannot be used.
 3. Target thread suddenly finds itself in Narnia (signal handler). Since the
    signal is the magical SIGUSR2, it proceeds to dump its stack frame addresses in
    a shared buffer. It then toggles completion, with release semantics, before
    returning to its regularly scheduled programming.
 4. Caller exits poll-loop and assembles a complete stack trace from the frame
    addresses in the shared buffer, demangled and shiny.

Signal-safe stack frame dumping currently uses Boost.Stacktrace, which is a
header-only dependency and appears to use libunwind under the hood.

### CODE REVIEW

**Comment by vekterli** [comment]
Hmm interesting, the expected function symbols in the test aren't visible when running in the CD build. Will investigate.

File: vespalib/src/vespa/vespalib/util/signalhandler.cpp [hunk 1]
```diff
@@ -116,23 +153,78 @@ SignalHandler::unhook()
...
+
+    assert(!_shared_backtrace_data._want_backtrace.load());
+    _shared_backtrace_data._signal_handler_done.store(false);
+
+    _shared_backtrace_data._want_backtrace.store(true);
+    if (pthread_kill(thread_id, SIGUSR2) != 0) {
+        _shared_backtrace_data._want_backtrace.store(false);
+        return "(pthread_kill() failed; could not get backtrace)";
+    }
+    while (!_shared_backtrace_data._signal_handler_done.load()) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/util/signalhandler.cpp:207]: consider using cmp exchange here as well to consume done signal

**Review by havardpe — APPROVED** [review]

File: vespalib/src/tests/signalhandler/signalhandler_test.cpp [hunk 2]
```diff
@@ -32,6 +32,60 @@ Test::Main()
...
+// phase latches it sort of feels like the semantics are more immediately obvious.
+void my_cool_function(std::latch& arrival_latch, std::latch& departure_latch) {
+    arrival_latch.arrive_and_wait();
+    // Twiddle thumbs in departure latch until main test thread has dumped our stack
+    departure_latch.arrive_and_wait();
+}
+
+TEST(SignalHandlerTest, can_dump_stack_of_another_thread)
+{
+    SignalHandler::enable_cross_thread_stack_tracing();
```
[the remark this answers is not in the document]
Comment (vekterli) [reply vespalib/src/tests/signalhandler/signalhandler_test.cpp]: Done

File: vespalib/src/vespa/vespalib/util/signalhandler.h [hunk 3]
```diff
@@ -43,27 +45,42 @@ class SignalHandler
      * dispatch the signal to the appropriate static signal handler
      * instance.
      *
+     * noinline to ensure consistent number of frames to skip for back-tracing.
+     *
      * @param signal the caught signal
      **/
-    static void handleSignal(int signal);
+    static void handleSignal(int signal) noexcept __attribute__((noinline));
```
[the remark this answers is not in the document]
Comment (vekterli) [reply vespalib/src/vespa/vespalib/util/signalhandler.h]: Fixed

File: vespalib/src/vespa/vespalib/util/signalhandler.cpp (hunk 1 above)
Comment (vekterli) [reply vespalib/src/vespa/vespalib/util/signalhandler.cpp]: Done

**Review by havardpe — APPROVED** [review]

**Pull request merged**